### ISSUE TITLE
[Bug]: EagerRuntime gradient accumulation has a read-modify-write race

### ISSUE THREAD

**Issue by terasakisatoshi** [tensor4all/tenferro-rs#1075]
## Summary
`EagerRuntime::store_grads` in `crates/tenferro-ad/src/eager.rs` releases the per-slot mutex after reading the current gradient, computes the accumulated value, and only re-acquires the lock later to write the result. Between the read and the write another thread can call `backward()` or `clear_grad()` on the same slot, causing lost updates or silently resurrecting a cleared gradient.

## Affected code
- `crates/tenferro-ad/src/eager.rs:718-757`

## Current behavior
```rust
for (slot, incoming) in updates {
    let next = {
        let current = slot.lock().unwrap(); // read
        match current.as_ref() { ... }
    }; // lock dropped
    staged.push((slot, next));
}
for (slot, next) in staged {
    *slot.lock().unwrap() = Some(next); // write later
}
```

## Reproducer sketch
```rust
use std::sync::Arc;
use std::thread;
use tenferro_ad::{eager_tensor, EagerRuntime, EagerTensor, Tensor};

let ctx = Arc::new(EagerRuntime::new());
let x = EagerTensor::from_tensor_in(
    Tensor::from_vec_col_major(vec![], vec![1.0_f64]),
    Arc::clone(&ctx),
);
x.set_requires_grad(true);

let loss = eager_tensor::mul(&x, &x).unwrap();

// Concurrent backward/clear calls can corrupt the gradient.
let handles: Vec<_> = (0..10)
    .map(|_| {
        let ctx = Arc::clone(&ctx);
        let x = x.clone();
        thread::spawn(move || {
            let _ = loss.backward(&ctx);
        })
    })
    .collect();
for h in handles { h.join().unwrap(); }

// The final gradient should be deterministic (2.0 for x^2 at x=1).
let grad = x.grad().unwrap();
```

## Expected behavior
Hold the per-slot mutex across the read-compute-write sequence, or use a compare-and-swap retry loop.

## Acceptance criteria
- [ ] Make gradient accumulation atomic with respect to concurrent `backward()`/`clear_grad()`.
- [ ] Regression test exercising concurrent gradient updates.
